### PR TITLE
Modularize try statement parsing

### LINKED ISSUE #375: catch statement causes syntax error
```javascript
try {
        var x = 1;
}
catch (e) { }
```
```
ParsingError: Expected token '(', got 'catch' in catch in try statement at line 4, col 1
```

### PR DESCRIPTION
**Pull request by abhijeetbhagat**
This Pull Request fixes/closes #375.

- Modularized the try statement parsing into separate modules as mentioned in the reply to the issue.
- Haven't implemented the entire catch parameter grammar parsing. Just kept the existing identifier parsing as it is for now.
- Added some tests for the try statement parsing.

### CODE REVIEW

**Comment by Razican** [comment]
Thanks for this contribution!
It seems that you need to re-base, since in #389 we added `BindingIdentifier` parsing. The change for this case was to parse the binding identifier like this:

```rust
// TODO: CatchParameter - BindingPattern
let catch_param = BindingIdentifier::new(self.allow_yield, self.allow_await)
    .parse(cursor)
    .map(Node::local)?;
```

File: boa/src/syntax/parser/statement/try_stm/mod.rs [hunk 1]
```diff
@@ -63,38 +69,24 @@ impl TokenParser for TryStatement {
...
-                Some(
-                    Block::new(self.allow_yield, self.allow_await, self.allow_return)
-                        .parse(cursor)?,
-                ),
-                Some(catch_param),
-            )
+            match Catch::new(self.allow_yield, self.allow_await, self.allow_return).parse(cursor) {
+                Ok((catch, param)) => (catch, param),
+                Err(e) => return Err(e),
+            }
```
Comment (Razican) [inline boa/src/syntax/parser/statement/try_stm/mod.rs:76]:
```suggestion
            Catch::new(self.allow_yield, self.allow_await, self.allow_return).parse(cursor)?
```

File: boa/src/syntax/parser/statement/try_stm/mod.rs [hunk 2]
```diff
@@ -63,38 +69,24 @@ impl TokenParser for TryStatement {
...
-            Some(Block::new(self.allow_yield, self.allow_await, self.allow_return).parse(cursor)?)
+        let next_token = cursor.peek(0);
+        let finally_block = if next_token.is_some()
+            && next_token.unwrap().kind == TokenKind::Keyword(Keyword::Finally)
+        {
+            match Finally::new(self.allow_yield, self.allow_await, self.allow_return).parse(cursor)
+            {
+                Ok(finally) => Some(finally),
+                Err(e) => return Err(e),
+            }
```
Comment (Razican) [inline boa/src/syntax/parser/statement/try_stm/mod.rs:89]:
```suggestion
            Finally::new(self.allow_yield, self.allow_await, self.allow_return).parse(cursor)?
```

File: boa/src/syntax/parser/statement/try_stm/mod.rs [hunk 3]
```diff
@@ -63,38 +69,24 @@ impl TokenParser for TryStatement {
...
         } else {
             (None, None)
         };
 
-        // FINALLY
-        let finally_block = if cursor.next_if(Keyword::Finally).is_some() {
-            Some(Block::new(self.allow_yield, self.allow_await, self.allow_return).parse(cursor)?)
+        let next_token = cursor.peek(0);
+        let finally_block = if next_token.is_some()
+            && next_token.unwrap().kind == TokenKind::Keyword(Keyword::Finally)
```
Comment (Razican) [inline boa/src/syntax/parser/statement/try_stm/mod.rs:83]: I would use a match statement to avoid `unwrap()` should be more efficient, too.

Comment (abhijeetbhagat) [reply boa/src/syntax/parser/statement/try_stm/mod.rs]: Done

File: boa/src/syntax/parser/statement/try_stm/mod.rs (hunk 2 above)
Comment (abhijeetbhagat) [reply boa/src/syntax/parser/statement/try_stm/mod.rs]: Done

**Pull request merged**